mitodl/mit-open
# UserList model and API

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #42 

#### What's this PR do?
- Adds `UserList` model, permissions, serializers, and views
- Puts some topic serializer functionality into a separate `WriteableTopicsMixin` class to avoid code duplication

#### How should this be manually tested?
- Go to the `http://localhost:8063/api/v1/userlists/` as an anonymous user, you should get a 403 response
- Log in (as anyone), go to the same url, and now you should be able to successfully post the following to create a new user list:
  ```
  {
    "topics": [1,2],
    "title": "My 1st UserList",
    "description": "My UL description"
  }
  ```
- Go to `http://localhost:8063/api/v1/userlists/<id>/` and PATCH this response, it should update the title.
  ```
  {
    "title": "My 1st UserList, modified",
  }
  ```
- Go to `http://localhost:8063/api/v1/userlists/<id>/resources/` and post the following, it should add the resource to the list.
  ```
  {
    "child": <id of a course or program learning resource>
  }
  ```

- Refresh `http://localhost:8063/api/v1/userlists/<id>/resources/`, you should see all the standard details of the added resource under `results`, along with a `position` of 0
- Add another, it should be added with a `position` of 1
- Go to `http://localhost:8063/api/v1/userlists/<user_list resource id>/resources/<user_list_relationship id>/` and click `DELETE`, the resource should be deleted from the userlist.
- Log in as another user and go to `http://localhost:8063/api/v1/userlists/`, you should not get any user lists returned.
- Go to `http://localhost:8063/api/v1/userlists/<id>`, you should get a 404
